### PR TITLE
Block Library: Add core template part block.

### PR DESCRIPTION
**Pull request by epiqueras**
Follows #18339

## Description

This PR adds the implementation for the `core/template-part` block introduced in #18339.

It focuses on the server-side render callback so that block templates that use template parts begin to render correctly on the front end. The editor implementation for creating and editing template parts will come in a follow-up PR.

## How has this been tested?

It was verified that the following steps make post previews render as expected:

Screen Shot 2019-11-25 at 4 11 13 PM

- Enable the Gutenberg full site editing experiment.
- Add the following files to your active theme's folder:

`/block-templates/single.html`:
```html
<!-- wp:template-part {"slug":"header","theme":"twentytwenty"} /-->
<!-- wp:paragraph -->
<p>Single Template</p>
<!-- /wp:paragraph -->
```

`/block-template-parts/header.html`:
```html
<!-- wp:paragraph -->
<p>Header TwentyTwenty Template Part</p>
<!-- /wp:paragraph -->
```

## Types of Changes

*New Feature:* Added a new server-rendered template part block for the full site editing experiment.

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. .

### CODE REVIEW

**Comment by ellatrix** [comment]
Looks like you still need test fixture files for this block.

**Comment by epiqueras** [comment]
You're right, thanks! I added them.

**Comment by ellatrix** [comment]
Why is there a `theme` attribute? Are these template parts meant to be used still when switching themes? You cannot have multiple themes active, so the attribute seem unnecessary to me.

**Comment by epiqueras** [comment]
From the PR that introduced it, #18339

>We need a `theme` attribute to differentiate between template parts with the same slug originating from different themes. When you save changes to a given template, it and all its nested template parts will persist across theme switches. Even though by that point those template parts will have ids and will be renderable regardless of what the new active theme provides, we need a way for the user to insert new instances of them and that is where `theme` comes in handy. For now, we store the value in post meta so we can do meta queries.

File: packages/block-library/src/template-part/index.php [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+		$content = get_post( $attributes['postId'] )->post_content;
+	} elseif ( wp_get_theme()->get( 'TextDomain' ) === $attributes['theme'] ) {
+		// Else, if the template part was provided by the active theme,
+		// render the corresponding file content.
+		$template_part_file_path =
+				get_stylesheet_directory() . '/block-template-parts/' . $attributes['slug'] . '.html';
+		if ( file_exists( $template_part_file_path ) ) {
+			$content = file_get_contents( $template_part_file_path );
+		}
+	}
```
Comment (youknowriad) [inline packages/block-library/src/template-part/index.php]: I'm pretty sure the logic here might need to evolve as we discover edge cases and we polish the flows. I'm not certain the current way of handing theme template parts is the best, but it's a decent one to start with.

**Review by youknowriad — APPROVED** [review]
LGTM

**Comment by youknowriad** [comment]
I managed to make this work (with a rebase) but while testing with https://github.com/carolinan/burgers I noticed this error 
Capture d’écran 2019-12-04 à 11 24 25 AM

File: packages/block-library/src/template-part/index.php (hunk 1 above)
Comment (epiqueras) [reply packages/block-library/src/template-part/index.php]: We'll need to handle recursive template parts differently, maybe.

**Comment by epiqueras** [comment]
What's in your line 158 after rebasing? It's probably just a missing if-guard.

**Pull request merged**